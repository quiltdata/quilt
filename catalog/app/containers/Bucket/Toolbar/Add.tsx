import * as React from 'react'

import { AddOutlined as IconAddOutlined } from '@material-ui/icons'

import * as Buttons from 'components/Buttons'

type ButtonProps = Omit<Buttons.WithPopoverProps, 'icon'>

export default function Button({ label = 'Add files', ...props }: ButtonProps) {
  return <Buttons.WithPopover icon={IconAddOutlined} label={label} {...props} />
}
