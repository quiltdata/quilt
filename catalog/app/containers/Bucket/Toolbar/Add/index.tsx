import * as React from 'react'

import { AddOutlined as IconAddOutlined } from '@material-ui/icons'

import * as Buttons from 'components/Buttons'

import type { DirHandle } from '../types'

import * as Context from './ContextDir'

export { default as BucketDirOptions } from './BucketDirOptions'
export { default as UploadDialog } from './UploadDialog'

interface ButtonProps {
  children: NonNullable<React.ReactNode>
  className?: string
  handle: DirHandle
  label?: string
}

export function Button({ children, handle, label = 'Add files', ...props }: ButtonProps) {
  return (
    <Context.Provider handle={handle}>
      <Buttons.WithPopover icon={IconAddOutlined} label={label} {...props}>
        {children}
      </Buttons.WithPopover>
    </Context.Provider>
  )
}
