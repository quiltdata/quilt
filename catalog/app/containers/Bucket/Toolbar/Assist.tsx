import * as React from 'react'
import * as M from '@material-ui/core'

import * as Assistant from 'components/Assistant'
import QuratorIcon from 'components/Assistant/UI/QuratorIcon'

interface AssistButtonProps extends M.IconButtonProps {
  title?: string
  message?: string
}

export default function AssistButton({ message, title, ...props }: AssistButtonProps) {
  const assist = Assistant.use()
  if (!assist) return null
  return (
    <M.IconButton
      color="primary"
      onClick={() => assist(message || 'Summarize this document')}
      style={{ marginTop: '-12px', marginBottom: '-12px' }}
      edge="end"
      {...props}
    >
      <M.Tooltip title={title || 'Summarize and chat with AI'}>
        <QuratorIcon />
      </M.Tooltip>
    </M.IconButton>
  )
}
