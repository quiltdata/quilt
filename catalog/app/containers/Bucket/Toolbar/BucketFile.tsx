import cx from 'classnames'
import * as React from 'react'
import * as M from '@material-ui/core'

import * as Buttons from 'components/Buttons'
import type { EditorState } from 'components/FileEditor'
import cfg from 'constants/config'
import * as BucketPreferences from 'utils/BucketPreferences'

import AssistButton from '../AssistButton'
import type { ViewModes } from '../viewModes'

import * as Get from './Get'
import * as Organize from './Organize'
import type { FileHandle } from './types'

interface Features {
  get: false | { code: boolean } | null
  organize: boolean | null
  qurator: boolean | null
}

function useFeatures(deleted?: boolean): Features | null {
  const { prefs } = BucketPreferences.use()
  return BucketPreferences.Result.match(
    {
      Ok: ({ ui: { actions, blocks } }) => ({
        get:
          !deleted && !cfg.noDownload && actions.downloadObject
            ? { code: blocks.code }
            : false,
        organize: true,
        qurator: blocks.qurator,
      }),
      _: () => null,
    },
    prefs,
  )
}

const useStyles = M.makeStyles((t) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
    gap: t.spacing(1),
  },
  qurator: {
    marginLeft: t.spacing(-1),
  },
}))

interface BucketFileProps {
  children?: React.ReactNode
  className?: string
  deleted?: boolean
  editorState: EditorState
  handle: FileHandle
  viewModes: ViewModes
  onReload: () => void
}

export default function BucketFile({
  children,
  className,
  deleted,
  editorState,
  handle,
  viewModes,
  onReload,
}: BucketFileProps) {
  const classes = useStyles()
  const features = useFeatures(deleted)

  if (!features)
    return (
      <div className={cx(classes.root, className)}>
        <Buttons.Skeleton size="small" />
        <Buttons.Skeleton size="small" />
        <Buttons.Skeleton size="small" />
      </div>
    )

  return (
    <div className={cx(classes.root, className)}>
      {children}

      {features.get && (
        <Get.Button label="Get file">
          <Get.BucketFileOptions handle={handle} hideCode={!features.get.code} />
        </Get.Button>
      )}

      {features.organize && (
        <Organize.Button onReload={onReload} handle={handle} editorState={editorState}>
          <Organize.BucketFileOptions viewModes={viewModes} />
        </Organize.Button>
      )}

      {features.qurator && <AssistButton className={classes.qurator} edge="end" />}
    </div>
  )
}
