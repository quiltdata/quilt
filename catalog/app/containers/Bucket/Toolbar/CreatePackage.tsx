import cx from 'classnames'
import * as React from 'react'
import * as M from '@material-ui/core'

import * as Buttons from 'components/Buttons'
import { useSelection } from 'containers/Bucket/Selection/Provider'

const useStyles = M.makeStyles({
  root: {
    transitionProperty: 'background-color, box-shadow', // Disable border-width transition
  },
})

interface ButtonProps {
  children: NonNullable<React.ReactNode>
  className?: string
  label?: string
}

export default function Button({ className, ...props }: ButtonProps) {
  const classes = useStyles()
  const slt = useSelection()
  return (
    <Buttons.WithPopover
      className={cx(classes.root, className)}
      label="Create package"
      variant={slt.isEmpty ? 'outlined' : 'contained'}
      color={slt.isEmpty ? 'default' : 'primary'}
      {...props}
    />
  )
}
