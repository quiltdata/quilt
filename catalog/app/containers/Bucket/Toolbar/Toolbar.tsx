import cx from 'classnames'
import * as React from 'react'
import * as M from '@material-ui/core'
import * as Icons from '@material-ui/icons'

import * as Buttons from 'components/Buttons'
import { useSelection } from 'containers/Bucket/Selection/Provider'

type ButtonProps = Omit<Buttons.WithPopoverProps, 'icon' | 'label'> &
  Partial<Pick<Buttons.WithPopoverProps, 'label'>>

export { default as Assist } from './Assist'

export function Add({ label = 'Add files', ...props }: ButtonProps) {
  return <Buttons.WithPopover icon={Icons.AddOutlined} label={label} {...props} />
}

export function Get({ label = 'Get files', ...props }: ButtonProps) {
  return <Buttons.WithPopover icon={Icons.GetAppOutlined} label={label} {...props} />
}

const useBadgeClasses = M.makeStyles({
  badge: {
    right: '4px',
  },
})

export function Organize({ label = 'Organize', ...props }: ButtonProps) {
  const slt = useSelection()
  const classes = useBadgeClasses()
  return (
    <M.Badge badgeContent={slt.totalCount} classes={classes} color="primary" max={999}>
      <Buttons.WithPopover
        icon={Icons.PlaylistAddCheckOutlined}
        label={label}
        disabled={slt.inited && !slt.totalCount}
        {...props}
      />
    </M.Badge>
  )
}

const useCreatePackageStyles = M.makeStyles({
  root: {
    transitionProperty: 'background-color, box-shadow',
  },
})

export function CreatePackage({
  label = 'Create package',
  className,
  ...props
}: ButtonProps) {
  const classes = useCreatePackageStyles()
  const slt = useSelection()
  return (
    <Buttons.WithPopover
      className={cx(classes.root, className)}
      label={label}
      variant={slt.isEmpty ? 'outlined' : 'contained'}
      color={slt.isEmpty ? 'default' : 'primary'}
      {...props}
    />
  )
}
