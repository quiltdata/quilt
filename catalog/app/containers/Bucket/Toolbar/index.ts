export * from './types'
export * from './Toolbar'
