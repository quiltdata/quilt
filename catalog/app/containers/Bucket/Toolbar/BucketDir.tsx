import cx from 'classnames'
import * as React from 'react'
import * as M from '@material-ui/core'

import * as Buttons from 'components/Buttons'
import cfg from 'constants/config'
import * as BucketPreferences from 'utils/BucketPreferences'

import * as Add from './Add'
import * as CreatePackage from './CreatePackage'
import * as Get from './Get'
import * as Organize from './Organize'
import type { DirHandle } from './types'

const useStyles = M.makeStyles((t) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
    gap: t.spacing(1),
  },
}))

interface Features {
  add: boolean | null
  get: false | { code: boolean } | null
  organize: boolean | null
  createPackage: boolean | null
}

function useFeatures(): Features | null {
  const { prefs } = BucketPreferences.use()
  return BucketPreferences.Result.match(
    {
      Ok: ({ ui: { actions, blocks } }) => ({
        add: actions.writeFile,
        get_code: blocks.code,
        get: !cfg.noDownload && actions.downloadObject ? { code: blocks.code } : false,
        organize: true,
        createPackage: actions.createPackage,
      }),
      _: () => null,
    },
    prefs,
  )
}

interface BucketDirProps {
  className?: string
  handle: DirHandle
}

export default function BucketDir({ className, handle }: BucketDirProps) {
  const classes = useStyles()
  const features = useFeatures()

  if (!features)
    return (
      <div className={cx(classes.root, className)}>
        <Buttons.Skeleton size="small" />
        <Buttons.Skeleton size="small" />
        <Buttons.Skeleton size="small" />
        <Buttons.Skeleton size="small" />
      </div>
    )

  return (
    <div className={cx(classes.root, className)}>
      {features.add && (
        <Add.Button>
          <Add.BucketDirOptions handle={handle} />
        </Add.Button>
      )}

      {features.get && (
        <Get.Button>
          <Get.BucketDirOptions handle={handle} hideCode={!features.get.code} />
        </Get.Button>
      )}

      {features.organize && (
        <Organize.Button>
          <Organize.BucketDirOptions handle={handle} />
        </Organize.Button>
      )}

      {features.createPackage && (
        <CreatePackage.Button handle={handle}>
          <CreatePackage.BucketDirOptions handle={handle} />
        </CreatePackage.Button>
      )}
    </div>
  )
}
