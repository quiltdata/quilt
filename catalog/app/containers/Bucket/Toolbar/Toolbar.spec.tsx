import * as React from 'react'
import { render, fireEvent, act } from '@testing-library/react'

import * as Toolbar from './Toolbar'

jest.mock('./Assist', jest.fn)

jest.mock('@material-ui/core', () => ({
  ...jest.requireActual('@material-ui/core'),
  IconButton: ({ className, children, onClick }: any) => (
    <button role="button" {...{ className, children, onClick }} />
  ),
  Button: ({ className, children, onClick }: any) => (
    <button role="button" {...{ className, children, onClick }} />
  ),
}))

const useSelection = jest.fn()
jest.mock('containers/Bucket/Selection/Provider', () => ({
  useSelection: () => useSelection(),
}))

describe('containers/Bucket/Toolbar', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  describe('Add', () => {
    it('should render with default label', () => {
      const { container } = render(
        <Toolbar.Add className="custom-class">Hello, Popover!</Toolbar.Add>,
      )
      expect(container).toMatchSnapshot()
    })

    it('should render with custom label', () => {
      const { container } = render(
        <Toolbar.Add label="Custom Add Label">Hello, Popover!</Toolbar.Add>,
      )
      expect(container).toMatchSnapshot()
    })

    it('should render children when popover is opened', () => {
      const { container, getByRole } = render(<Toolbar.Add>Hello, Popover!</Toolbar.Add>)
      const button = getByRole('button')
      act(() => {
        fireEvent.click(button)
      })
      expect(container).toMatchSnapshot()
    })
  })

  describe('Get', () => {
    it('should render with default label', () => {
      const { container } = render(
        <Toolbar.Get className="custom-class">Hello, Popover!</Toolbar.Get>,
      )
      expect(container).toMatchSnapshot()
    })

    it('should render with custom label', () => {
      const { container } = render(
        <Toolbar.Get label="Custom Get Label">Hello, Popover!</Toolbar.Get>,
      )
      expect(container).toMatchSnapshot()
    })

    it('should render children when popover is opened', () => {
      const { container, getByRole } = render(<Toolbar.Get>Hello, Popover!</Toolbar.Get>)
      const button = getByRole('button')
      act(() => {
        fireEvent.click(button)
      })
      expect(container).toMatchSnapshot()
    })
  })

  describe('CreatePackage', () => {
    beforeEach(() => {
      useSelection.mockReturnValue({
        isEmpty: true,
        totalCount: 0,
        inited: true,
      })
    })

    it('should render with empty selection', () => {
      const { container } = render(
        <Toolbar.CreatePackage className="custom-class">
          Hello, Popover!
        </Toolbar.CreatePackage>,
      )
      expect(container).toMatchSnapshot()
    })

    it('should render children when popover is opened', () => {
      const { container, getByRole } = render(
        <Toolbar.CreatePackage>Hello, Popover!</Toolbar.CreatePackage>,
      )
      const button = getByRole('button')
      act(() => {
        fireEvent.click(button)
      })
      expect(container).toMatchSnapshot()
    })
  })

  describe('Organize', () => {
    beforeEach(() => {
      useSelection.mockReturnValue({
        isEmpty: false,
        totalCount: 5,
        inited: true,
      })
    })

    it('should render with selection items', () => {
      const { container } = render(
        <Toolbar.Organize className="custom-class">Hello, Popover!</Toolbar.Organize>,
      )
      expect(container).toMatchSnapshot()
    })

    it('should render children when popover is opened', () => {
      const { container, getByRole } = render(
        <Toolbar.Organize>Hello, Popover!</Toolbar.Organize>,
      )
      const button = getByRole('button')
      act(() => {
        fireEvent.click(button)
      })
      expect(container).toMatchSnapshot()
    })
  })
})
