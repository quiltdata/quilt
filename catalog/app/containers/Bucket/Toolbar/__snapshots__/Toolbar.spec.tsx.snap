// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Bucket/Toolbar Add should render children when popover is opened 1`] = `
<div>
  <div
    class="makeStyles-root-9"
  >
    <button
      role="button"
    >
      Add files
    </button>
    <div
      aria-hidden="true"
      class="MuiBackdrop-root makeStyles-backdrop-11"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      class="MuiPaper-root makeStyles-popup-10 MuiPaper-elevation4 MuiPaper-rounded"
    >
      Hello, Popover!
    </div>
  </div>
</div>
`;

exports[`containers/Bucket/Toolbar Add should render with custom label 1`] = `
<div>
  <div
    class="makeStyles-root-5"
  >
    <button
      role="button"
    >
      Custom Add Label
    </button>
    <div
      aria-hidden="true"
      class="MuiBackdrop-root makeStyles-backdrop-7"
      style="opacity: 0; visibility: hidden;"
    />
  </div>
</div>
`;

exports[`containers/Bucket/Toolbar Add should render with default label 1`] = `
<div>
  <div
    class="makeStyles-root-1"
  >
    <button
      class="custom-class"
      role="button"
    >
      Add files
    </button>
    <div
      aria-hidden="true"
      class="MuiBackdrop-root makeStyles-backdrop-3"
      style="opacity: 0; visibility: hidden;"
    />
  </div>
</div>
`;

exports[`containers/Bucket/Toolbar CreatePackage should render children when popover is opened 1`] = `
<div>
  <div
    class="makeStyles-root-31"
  >
    <button
      class="makeStyles-root-30"
      role="button"
    >
      Create package
    </button>
    <div
      aria-hidden="true"
      class="MuiBackdrop-root makeStyles-backdrop-33"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      class="MuiPaper-root makeStyles-popup-32 MuiPaper-elevation4 MuiPaper-rounded"
    >
      Hello, Popover!
    </div>
  </div>
</div>
`;

exports[`containers/Bucket/Toolbar CreatePackage should render with empty selection 1`] = `
<div>
  <div
    class="makeStyles-root-26"
  >
    <button
      class="makeStyles-root-25 custom-class"
      role="button"
    >
      Create package
    </button>
    <div
      aria-hidden="true"
      class="MuiBackdrop-root makeStyles-backdrop-28"
      style="opacity: 0; visibility: hidden;"
    />
  </div>
</div>
`;

exports[`containers/Bucket/Toolbar Get should render children when popover is opened 1`] = `
<div>
  <div
    class="makeStyles-root-21"
  >
    <button
      role="button"
    >
      Get files
    </button>
    <div
      aria-hidden="true"
      class="MuiBackdrop-root makeStyles-backdrop-23"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      class="MuiPaper-root makeStyles-popup-22 MuiPaper-elevation4 MuiPaper-rounded"
    >
      Hello, Popover!
    </div>
  </div>
</div>
`;

exports[`containers/Bucket/Toolbar Get should render with custom label 1`] = `
<div>
  <div
    class="makeStyles-root-17"
  >
    <button
      role="button"
    >
      Custom Get Label
    </button>
    <div
      aria-hidden="true"
      class="MuiBackdrop-root makeStyles-backdrop-19"
      style="opacity: 0; visibility: hidden;"
    />
  </div>
</div>
`;

exports[`containers/Bucket/Toolbar Get should render with default label 1`] = `
<div>
  <div
    class="makeStyles-root-13"
  >
    <button
      class="custom-class"
      role="button"
    >
      Get files
    </button>
    <div
      aria-hidden="true"
      class="MuiBackdrop-root makeStyles-backdrop-15"
      style="opacity: 0; visibility: hidden;"
    />
  </div>
</div>
`;

exports[`containers/Bucket/Toolbar Organize should render children when popover is opened 1`] = `
<div>
  <span
    class="MuiBadge-root"
  >
    <div
      class="makeStyles-root-41"
    >
      <button
        role="button"
      >
        Organize
      </button>
      <div
        aria-hidden="true"
        class="MuiBackdrop-root makeStyles-backdrop-43"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        class="MuiPaper-root makeStyles-popup-42 MuiPaper-elevation4 MuiPaper-rounded"
      >
        Hello, Popover!
      </div>
    </div>
    <span
      class="MuiBadge-badge makeStyles-badge-40 MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorPrimary"
    >
      5
    </span>
  </span>
</div>
`;

exports[`containers/Bucket/Toolbar Organize should render with selection items 1`] = `
<div>
  <span
    class="MuiBadge-root"
  >
    <div
      class="makeStyles-root-36"
    >
      <button
        class="custom-class"
        role="button"
      >
        Organize
      </button>
      <div
        aria-hidden="true"
        class="MuiBackdrop-root makeStyles-backdrop-38"
        style="opacity: 0; visibility: hidden;"
      />
    </div>
    <span
      class="MuiBadge-badge makeStyles-badge-35 MuiBadge-anchorOriginTopRightRectangle MuiBadge-colorPrimary"
    >
      5
    </span>
  </span>
</div>
`;
